File: iq/11/geomseq/find_test.go
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package geomseq_test

import (
	"testing"

	"github.com/google/go-cmp/cmp"
	"github.com/skhal/lab/iq/11/geomseq"
)

func TestFind(t *testing.T) {
	tests := []struct {
		name string
		nn   []int
		r    geomseq.Ratio
		want []geomseq.Triplet
	}{
		{
			name: "empty",
			r:    2,
		},
		{
			name: "one item",
			nn:   []int{1},
			r:    2,
		},
		{
			name: "two items",
			nn:   []int{1, 2},
			r:    2,
		},
		{
			name: "thee items no seq",
			nn:   []int{1, 2, 3},
			r:    2,
		},
		{
			name: "thee items seq",
			nn:   []int{1, 2, 4},
			r:    2,
			want: []geomseq.Triplet{
				{0, 1, 2},
			},
		},
		{
			name: "four items no seq",
			nn:   []int{1, 2, 3, 5},
			r:    2,
		},
		{
			name: "four items two seq",
			nn:   []int{1, 2, 2, 4},
			r:    2,
			want: []geomseq.Triplet{
				{0, 1, 3},
				{0, 2, 3},
			},
		},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			got := geomseq.Find(tc.nn, tc.r)

			if diff := cmp.Diff(tc.want, got); diff != "" {
				t.Errorf("geomseq.Find(%v, %d) mismatch (-want, +got):\n%s", tc.nn, tc.r, diff)
			}
		})
	}
}

File: iq/11/geomseq/find_test.cc
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

//
// NOLINTNEXTLINE
//go:build ignore

#include "iq/11/geomseq/find.h"

#include <cctype>
#include <iomanip>
#include <ostream>
#include <sstream>
#include <string>
#include <string_view>
#include <tuple>
#include <vector>

#include "absl/strings/str_format.h"
#include "absl/strings/str_join.h"
#include "gmock/gmock.h"
#include "gtest/gtest.h"

namespace iq::geomseq {

// Must be in the same namespace with Triplet:
// https://google.github.io/googletest/advanced.html#teaching-googletest-how-to-print-your-values
template <typename Sink>
void AbslStringify(Sink& sink, const Triplet& t) {
  std::ostringstream os;
  os << t;
  absl::Format(&sink, os.str());
}

namespace {

using ::testing::Pointwise;
using ::testing::TestParamInfo;
using ::testing::TestWithParam;
using ::testing::ValuesIn;

constexpr std::string_view kStrSpace = " ";
constexpr char kCharUnderscore = '_';

struct FindTestParam {
  std::string name;
  std::vector<int> nn = {};
  int ratio;
  std::vector<Triplet> want = {};

  friend std::ostream& operator<<(std::ostream& os, const FindTestParam& tp) {
    return os << "name: " << std::quoted(tp.name) << " nn: ["
              << absl::StrJoin(tp.nn, kStrSpace) << "]";
  }
};

class FindTest : public TestWithParam<FindTestParam> {};

MATCHER(TripletEq, "equal triplets") {
  const Triplet& got = std::get<0>(arg);
  const Triplet& want = std::get<1>(arg);
  return std::tie(got.i, got.j, got.k) == std::tie(want.i, want.j, want.k);
}

TEST_P(FindTest, Test) {
  const FindTestParam& tp = GetParam();

  const std::vector<Triplet> got = Find(tp.nn, tp.ratio);

  EXPECT_THAT(got, Pointwise(TripletEq(), tp.want));
}

const FindTestParam kFindTestParams[]{
    {.name = "empty", .ratio = 2},
    {.name = "one item", .nn = {1}, .ratio = 2},
    {.name = "two items", .nn = {1, 2}, .ratio = 2},
    {.name = "three items no seq", .nn = {1, 2, 3}, .ratio = 2},
    {
        .name = "three items seq",
        .nn = {1, 2, 4},
        .ratio = 2,
        .want =
            {
                {0, 1, 2},
            },
    },
    {
        .name = "four items two seq",
        .nn = {1, 2, 2, 4},
        .ratio = 2,
        .want =
            {
                {0, 1, 3},
                {0, 2, 3},
            },
    },
};

INSTANTIATE_TEST_SUITE_P(FindTest, FindTest, ValuesIn(kFindTestParams),
                         [](const TestParamInfo<FindTestParam>& info) {
                           std::string name = info.param.name;
                           std::replace_if(
                               name.begin(), name.end(),
                               [](char c) { return !std::isalnum(c); },
                               kCharUnderscore);
                           return name;
                         });

}  // namespace
}  // namespace iq::geomseq
